rust_src/src/casetab.rs

```diff
@@ -3,35 +3,200 @@
+                canon = make_char_table(Qcase_table.into(), Qnil).into();
+                self.set_extras(1, canon);
+                map_char_table(Some(set_canon), Qnil, table, table);
+            }
+
+            if eqv.is_nil() {
+                eqv = make_char_table(Qcase_table.into(), Qnil).into();
+                map_char_table(Some(set_identity), Qnil, canon, eqv);
+                map_char_table(Some(shuffle), Qnil, canon, eqv);
+                self.set_extras(2, eqv);
+            }
+
+            // This is so set_image_of_range_1 can find the EQV table.
+            LispCaseTable::from(canon).set_extras(2, eqv);
+        }
+    }
+
+    pub fn get(self, idx: isize) -> LispObject {
+        self.0.get(idx)
+    }
+
+    pub fn extras(self) -> (LispObject, LispObject, LispObject) {
+        let extras = unsafe { self.0.extras.as_slice(3) };
+        (extras[0], extras[1], extras[2])
+    }
+
+    pub fn set_extras(&mut self, idx: isize, value: LispObject) {
+        assert!(0 <= idx && idx < 3);
+        let extras = unsafe { self.0.extras.as_mut_slice(3) };
+        extras[idx as usize] = value;
+    }
+}
+
+impl From<LispObject> for Option<LispCaseTable> {
+    fn from(obj: LispObject) -> Self {
+        let table = match Option::<LispCharTableRef>::from(obj) {
+            None => {
+                return None;
+            }
+            Some(t) => {
+                if t.purpose.eq(Qcase_table) {
+                    LispCaseTable(t)
+                } else {
+                    return None;
+                }
+            }
+        };
+
+        let (up, canon, eqv) = table.extras();
+
+        if (up.is_nil() || up.is_char_table())
+            && ((canon.is_nil() && eqv.is_nil())
+                || (canon.is_char_table() && (eqv.is_nil() || eqv.is_char_table())))
+        {
+            Some(table)
+        } else {
+            None
+        }
+    }
+}
+
+impl From<LispObject> for LispCaseTable {
+    fn from(obj: LispObject) -> Self {
+        Option::<Self>::from(obj).unwrap_or_else(|| wrong_type!(Qcase_table_p, obj))
+    }
+}
+
+impl From<LispCaseTable> for LispObject {
+    fn from(table: LispCaseTable) -> Self {
+        table.0.into()
+    }
+}
+
+static mut Vascii_downcase_table: LispObject = Qnil;
+static mut Vascii_upcase_table: LispObject = Qnil;
+static mut Vascii_canon_table: LispObject = Qnil;
+static mut Vascii_eqv_table: LispObject = Qnil;
+
+// The following functions are called in map_char_table.
+
+// Set CANON char-table element for characters in RANGE to a
+// translated ELT by UP and DOWN char-tables.  This is done only when
+// ELT is a character.  The char-tables CANON, UP, and DOWN are in
+// TABLE.
+extern "C" fn set_canon(table: LispObject, range: LispObject, elt: LispObject) {
+    if let Some(idx) = elt.as_natnum() {
+        let case_table: LispCaseTable = table.into();
+
+        let (up, canon, _) = case_table.extras();
+
+        let up_table: LispCharTableRef = up.into();
```

Comment:
Is this conversion guaranteed? It looks like `.extras()` can also return nil objects.

Reply:
Hmmm. This is run to set the values.

This is the current version in Emacs:

```
set_canon:
  Lisp_Object up = XCHAR_TABLE (case_table)->extras[0];
  Lisp_Object canon = XCHAR_TABLE (case_table)->extras[1];

  if (FIXNATP (elt))
    Fset_char_table_range (canon, range, Faref (case_table, Faref (up, elt)));
```

If `up` is nil, that call to `Faref` will throw `wrong_type(Qarrayp)`. What that means is my code likely changes the error message. But now it is a useful error instead of the message from `aref`.

If we want to update this we should do it on my previous PR, which I based this one on.

Reply:
I would like to land the casetab.c PR and rebase this one. I posted it as a place holder.